networkx/algorithms/approximation/traveling_salesman.py

```diff
@@ -321,7 +322,14 @@ def traveling_salesman_problem(G, weight="weight", nodes=None, cycle=True, metho
             if u == v:
                 continue
             GG.add_edge(u, v, weight=dist[u][v])
-    best_GG = method(GG, weight)
+
+    # Several methods using init_cycle is a positional argument.
+    # To avoid chaning the API for those methods, we use this check.
+    if "init_cycle" in kwargs:
+        init_cycle = kwargs.pop("init_cycle")
+        best_GG = method(GG, init_cycle, **kwargs)
+    else:
+        best_GG = method(GG, weight, **kwargs)
```

Comment:
presumably those places using `init_cycle` are internal to networkx. Could you give an example of what can't be changed to a kwarg?

Reply:
These functions are exposed to library users as part of the public API, namely [`simulated_annealing_tsp`](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.approximation.traveling_salesman.simulated_annealing_tsp.html) and [`threshold_accepting_tsp`](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.approximation.traveling_salesman.threshold_accepting_tsp.html) and the usage of `init_cycle` can't be changed to a keyword argument since no default value is provided. We've discussed setting the `"greedy"` special value as the default, but there is no guarantee that that initial cycle leads to a good TSP approximation and we hope that not setting a default value will prompt users to think about if there might be a better initial cycle to seed the algorithms with.

Reply:
But why does that mean that it can't be supplied as a keyword argument?
```python
import networkx as nx
G = nx.complete_graph(4)
nx.algorithms.approximation.simulated_annealing_tsp(G, init_cycle="greedy")
```
works fine

Reply:
True, but in 

```python
best_GG = method(GG, weight, **kwargs)
```

The interpreter will pass in `weight` as `init_cycle` and then also find `init_cycle` in the `kwargs` dict and not know how to reconcile the two different values. 

However, your thoughts got me thinking a little harder about this, and I realized that there is a better solution, one which doesn't require changing any API. We can write 

```python
best_GG = method(GG, weight=weight, **kwargs)
```

and now even if `init_cycle` is in `kwargs` it will not be matched to the second positional argument since everything is using keywords (except `G`).

Reply:
This is better than removing `weight="weight"` from the `traveling_salesman_problem` method signature since it doesn't break any code which is passing the `weight` parameter by position and not keyword.

Reply:
This seems much cleaner!